cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/MinTopicLeadersPerBrokerGoal.java

```diff
@@ -0,0 +1,441 @@
+          leaderMoved = true;
+          break; // Successfully move one leader replica
+        }
+      }
+      if (leaderMoved) {
+        topicLeaderCountOnReceiverBroker++;
+        if (topicLeaderCountOnReceiverBroker >= minTopicLeadersPerBroker()) {
+          return; // This broker satisfies this goal for the given topic
+        }
+        topicLeaderCountOnGiverBroker--;
+        if (topicLeaderCountOnGiverBroker > minTopicLeadersPerBroker()) {
+          // Still have excessive topic leader to give
+          brokersWithExcessiveLeaderToMove.add(brokerWithExcessiveLeaderToMove);
+        }
+      }
+    }
+    throw new OptimizationFailureException(String.format("Cannot make broker %d have at least %d leader replica(s) of topic %s.",
+                                                         broker.id(),
+                                                         minTopicLeadersPerBroker(),
+                                                         topicMustHaveLeaderPerBroker));
+  }
+
+  /**
+   * This method creates a priority queue which has the broker with the most number of leader replicas of a given topic at the top of queue
+   * @param topicName name of the given topic
+   * @param clusterModel cluster model
+   * @param originalBroker original broker which needs to be excluded from the priority queue
+   * @return a priority queue which has the broker with the most number of leader replicas of a given topic at the top of queue
+   */
+  private PriorityQueue<Broker> getBrokersWithExcessiveLeaderToMove(String topicName, ClusterModel clusterModel, Broker originalBroker) {
+    PriorityQueue<Broker> brokersWithExcessiveLeaderToMove = new PriorityQueue<>((broker1, broker2) -> {
+      int broker1LeaderCount = broker1.numLeadersFor(topicName);
+      int broker2LeaderCount = broker2.numLeadersFor(topicName);
+      int leaderCountCompareResult = Integer.compare(broker2LeaderCount, broker1LeaderCount);
+      return leaderCountCompareResult == 0 ? Integer.compare(broker1.id(), broker2.id()) : leaderCountCompareResult;
+    });
+    clusterModel.aliveBrokers()
+                .stream()
+                .filter(broker -> broker.numLeadersFor(topicName) > minTopicLeadersPerBroker())
+                .forEach(brokersWithExcessiveLeaderToMove::add);
+    return brokersWithExcessiveLeaderToMove;
+  }
+
+  private Set<Broker> eligibleBrokersForLeadership(ClusterModel clusterModel, OptimizationOptions optimizationOptions) {
+    return clusterModel.aliveBrokers()
+                       .stream()
+                       .filter(broker -> isEligibleToHaveLeaders(broker, optimizationOptions))
+                       .collect(Collectors.toSet());
+  }
+
+  private boolean isEligibleToHaveLeaders(Broker broker, OptimizationOptions optimizationOptions) {
+    return !optimizationOptions.excludedBrokersForLeadership().contains(broker.id())
+           && !optimizationOptions.excludedBrokersForReplicaMove().contains(broker.id());
+  }
```

Comment:
Nit: Can we make (1) eligibleBrokersForLeadership and (2) isEligibleToHaveLeaders functions `static`?